### ISSUE TITLE
inconsistent flagging of SC2178/SC2128 based on `local` usage

### ISSUE THREAD

**Issue by vapier** [koalaman/shellcheck#1309]
#### For bugs
- Rule Id: SC2178/SC2128
- My shellcheck version: online

#### Here's a snippet or screenshot that shows the problem:

if we start with this code:
```sh
#!/bin/bash
get_vars() {
  local vars=( 1 2 3 )
  echo "${vars[@]}"
}
main() {
  local vars="$(get_vars)"
  echo "${vars}"
}
```

we get:
```sh
$ shellcheck myscript
 
Line 7:
  local vars="$(get_vars)"
        ^-- SC2155: Declare and assign separately to avoid masking return values.
        ^-- SC2178: Variable was used as an array but is now assigned a string.
 
Line 8:
  echo "${vars}"
        ^-- SC2128: Expanding an array without an index only gives the first element.
```

but if we fix SC2155, all the other warnings disappear:
```sh
#!/bin/bash
get_vars() {
  local vars=( 1 2 3 )
  echo "${vars[@]}"
}
main() {
  local vars
  vars="$(get_vars)"
  echo "${vars}"
}
```

```sh
$ shellcheck myscript
No issues detected!
```

if we remove the `local` decl, they come back:
```sh
#!/bin/bash
get_vars() {
  local vars=( 1 2 3 )
  echo "${vars[@]}"
}
main() {
  vars="$(get_vars)"
  echo "${vars}"
}
```
```sh
$ shellcheck myscript
 
Line 7:
  vars="$(get_vars)"
  ^-- SC2178: Variable was used as an array but is now assigned a string.
 
Line 8:
  echo "${vars}"
        ^-- SC2128: Expanding an array without an index only gives the first element.
```

this is weird because the scope of var from `main` to `get_vars` hasn't changed ... the `local var` in `main` still cascades down into `get_var`.

the SC2178/SC2128 should always be flagged even if the parent func declared it `local` for its own usage.

**Comment by ngzhian** [comment]
This is probably because ShellCheck does not have proper control flow, similar to #1225

**Comment by kolyshkin** [comment]
Similarly, there are cases when SC2128 gets misdetected because a variable comes from a caller's context. For example:

```bash
#!/bin/bash

x() {
	# Here we have args that came from y, i.e. args is an array.
	# This is clearly SC2128 ("Expanding an array without an index ..."),
	# but it is not detected.
	echo x: "$args"

	# Redeclare args as a string, shadowing the parent one.
	local args="three four"
	echo x: "$args"
}

y() {
	# Here we have args that came from global context, i.e. args is an array.
        # This is clearly SC2128 ("Expanding an array without an index ..."),
	# but it is not detected.
	echo y: "$args"

	# Redeclare args.
	local args=(one two)
	echo y: "${args[@]}"
	x
	echo y: "${args[@]}"
}

args=(global args)
y
```

**Comment by MatthijsBurgh** [comment]
Even with all local declarations on a separate line, I still get this issue when running checking multiple files at once. I only get the `SC2128` flag. The `SC2127` did show up when declaring and assigning on the same line, but that one disappeared. (GH actions run: https://github.com/tue-robotics/tue-env/actions/runs/3648670225/jobs/6162361889)
